File: sql-bulk-insert-benchmark/ruby/benchmark_mysql.rb
#!/usr/bin/env ruby
# MySQL用Rubyベンチマーク

require 'active_record'
require 'activerecord-import'
require 'mysql2'
require_relative 'data_generator'
require_relative 'benchmark_utils'

# MySQL接続設定
ActiveRecord::Base.establish_connection(
  adapter: 'mysql2',
  host: ENV.fetch('MYSQL_HOST', 'localhost'),
  port: ENV.fetch('MYSQL_PORT', 3306).to_i,
  username: ENV.fetch('MYSQL_USER', 'benchmark'),
  password: ENV.fetch('MYSQL_PASSWORD', 'benchmark'),
  database: ENV.fetch('MYSQL_DB', 'benchmark'),
  pool: 5
)

# モデル定義
class SimpleRecord < ActiveRecord::Base
  self.table_name = 'simple_records'
end

def cleanup_table
  SimpleRecord.connection.execute("TRUNCATE TABLE simple_records")
end

# ========================================
# Benchmark 1: create (1件ずつ)
# ========================================
def bench_create(record_count)
  cleanup_table
  records = DataGenerator.generate_simple_records(record_count)

  records.each do |record|
    SimpleRecord.create(record)
  end
end

# ========================================
# Benchmark 2: insert_all
# ========================================
def bench_insert_all(record_count)
  cleanup_table
  records = DataGenerator.generate_simple_records(record_count)
  SimpleRecord.insert_all(records)
end

# ========================================
# Benchmark 3: insert_all with batching
# ========================================
def bench_insert_all_batched(record_count, batch_size: 1000)
  cleanup_table
  records = DataGenerator.generate_simple_records(record_count)

  records.each_slice(batch_size) do |batch|
    SimpleRecord.insert_all(batch)
  end
end

# ========================================
# Benchmark 4: activerecord-import
# ========================================
def bench_activerecord_import(record_count)
  cleanup_table
  records = DataGenerator.generate_simple_records(record_count)

  objects = records.map { |r| SimpleRecord.new(r) }
  SimpleRecord.import(objects)
end

# ========================================
# Benchmark 5: activerecord-import with batching
# ========================================
def bench_activerecord_import_batched(record_count, batch_size: 1000)
  cleanup_table
  records = DataGenerator.generate_simple_records(record_count)

  records.each_slice(batch_size) do |batch|
    objects = batch.map { |r| SimpleRecord.new(r) }
    SimpleRecord.import(objects)
  end
end

# ========================================
# Benchmark 6: activerecord-import (配列形式)
# ========================================
def bench_activerecord_import_arrays(record_count)
  cleanup_table
  records = DataGenerator.generate_simple_records(record_count)

  columns = [:name, :email, :age, :score]
  values = records.map { |r| [r[:name], r[:email], r[:age], r[:score]] }

  SimpleRecord.import(columns, values)
end

# ========================================
# Benchmark 7: Raw mysql2 gem
# ========================================
def bench_raw_mysql2(record_count, batch_size: 1000)
  cleanup_table
  records = DataGenerator.generate_simple_records(record_count)

  client = Mysql2::Client.new(
    host: ENV.fetch('MYSQL_HOST', 'localhost'),
    port: ENV.fetch('MYSQL_PORT', 3306).to_i,
    username: ENV.fetch('MYSQL_USER', 'benchmark'),
    password: ENV.fetch('MYSQL_PASSWORD', 'benchmark'),
    database: ENV.fetch('MYSQL_DB', 'benchmark')
  )

  records.each_slice(batch_size) do |batch|
    values = batch.map do |r|
      "(#{client.escape(r[:name].to_s)}, #{client.escape(r[:email].to_s)}, #{r[:age]}, #{r[:score]})"
    end.join(', ')

    sql = "INSERT INTO simple_records (name, email, age, score) VALUES #{values}"
    client.query(sql)
  end

  client.close
end

# ========================================
# Benchmark 8: Raw mysql2 with prepared statement
# ========================================
def bench_raw_mysql2_prepared(record_count)
  cleanup_table
  records = DataGenerator.generate_simple_records(record_count)

  client = Mysql2::Client.new(
    host: ENV.fetch('MYSQL_HOST', 'localhost'),
    port: ENV.fetch('MYSQL_PORT', 3306).to_i,
    username: ENV.fetch('MYSQL_USER', 'benchmark'),
    password: ENV.fetch('MYSQL_PASSWORD', 'benchmark'),
    database: ENV.fetch('MYSQL_DB', 'benchmark')
  )

  stmt = client.prepare("INSERT INTO simple_records (name, email, age, score) VALUES (?, ?, ?, ?)")

  records.each do |r|
    stmt.execute(r[:name], r[:email], r[:age], r[:score])
  end

  stmt.close
  client.close
end

# ========================================
# Benchmark 9: Multi-row INSERT with mysql2
# ========================================
def bench_mysql2_multi_row(record_count, batch_size: 1000)
  cleanup_table
  records = DataGenerator.generate_simple_records(record_count)

  client = Mysql2::Client.new(
    host: ENV.fetch('MYSQL_HOST', 'localhost'),
    port: ENV.fetch('MYSQL_PORT', 3306).to_i,
    username: ENV.fetch('MYSQL_USER', 'benchmark'),
    password: ENV.fetch('MYSQL_PASSWORD', 'benchmark'),
    database: ENV.fetch('MYSQL_DB', 'benchmark')
  )

  records.each_slice(batch_size) do |batch|
    placeholders = (['(?, ?, ?, ?)'] * batch.size).join(', ')
    sql = "INSERT INTO simple_records (name, email, age, score) VALUES #{placeholders}"

    stmt = client.prepare(sql)
    values = batch.flat_map { |r| [r[:name], r[:email], r[:age], r[:score]] }
    stmt.execute(*values)
    stmt.close
  end

  client.close
end

# メイン実行
def main
  runner = BenchmarkRunner.new(output_file: '/results/ruby_mysql_results.json')

  test_sizes = [
    100_000,
    1_000_000,
    # 10_000_000,
  ]

  test_sizes.each do |size|
    puts "\n#{'#'*60}"
    puts "# Testing with #{size.to_s.reverse.gsub(/(\d{3})(?=\d)/, '\\1,').reverse} records"
    puts "#{'#'*60}"

    # createは遅すぎるので10万件のみ
    if size <= 100_000
      runner.run_benchmark("mysql_create_#{size}", size) do
        bench_create(size)
      end

      runner.run_benchmark("mysql_raw_prepared_#{size}", size) do
        bench_raw_mysql2_prepared(size)
      end
    end

    runner.run_benchmark("mysql_insert_all_#{size}", size) do
      bench_insert_all(size)
    end

    runner.run_benchmark("mysql_insert_all_batched_#{size}", size) do
      bench_insert_all_batched(size)
    end

    runner.run_benchmark("mysql_activerecord_import_#{size}", size) do
      bench_activerecord_import(size)
    end

    runner.run_benchmark("mysql_activerecord_import_batched_#{size}", size) do
      bench_activerecord_import_batched(size)
    end

    runner.run_benchmark("mysql_activerecord_import_arrays_#{size}", size) do
      bench_activerecord_import_arrays(size)
    end

    runner.run_benchmark("mysql_raw_mysql2_#{size}", size) do
      bench_raw_mysql2(size)
    end

    runner.run_benchmark("mysql_mysql2_multi_row_#{size}", size) do
      bench_mysql2_multi_row(size)
    end
  end

  runner.save_results
  runner.print_summary
end

if __FILE__ == $0
  puts "MySQL Bulk Insert Benchmark - Ruby"
  puts "=" * 60
  main
end

File: sql-bulk-insert-benchmark/python/benchmark_mysql.py
"""
MySQL用Pythonベンチマーク
"""
import os
import pymysql
import MySQLdb
from data_generator import DataGenerator
from benchmark_utils import BenchmarkRunner


# 接続設定
DB_CONFIG = {
    'host': os.getenv('MYSQL_HOST', 'localhost'),
    'port': int(os.getenv('MYSQL_PORT', 3306)),
    'user': os.getenv('MYSQL_USER', 'benchmark'),
    'password': os.getenv('MYSQL_PASSWORD', 'benchmark'),
    'database': os.getenv('MYSQL_DB', 'benchmark')
}


def get_pymysql_connection():
    """PyMySQL接続を取得"""
    return pymysql.connect(**DB_CONFIG)


def get_mysqldb_connection():
    """MySQLdb接続を取得"""
    return MySQLdb.connect(**DB_CONFIG)


def cleanup_table(table_name: str = 'simple_records'):
    """テーブルをクリーンアップ"""
    conn = get_pymysql_connection()
    cur = conn.cursor()
    cur.execute(f"TRUNCATE TABLE {table_name}")
    conn.commit()
    cur.close()
    conn.close()


# ========================================
# PyMySQL Benchmarks
# ========================================

def bench_pymysql_execute(record_count: int):
    """PyMySQL: 1件ずつexecute（遅い）"""
    cleanup_table()
    conn = get_pymysql_connection()
    cur = conn.cursor()

    records = DataGenerator.generate_simple_tuples(record_count)
    sql = "INSERT INTO simple_records (name, email, age, score) VALUES (%s, %s, %s, %s)"

    for record in records:
        cur.execute(sql, record)

    conn.commit()
    cur.close()
    conn.close()


def bench_pymysql_executemany(record_count: int):
    """PyMySQL: executemany"""
    cleanup_table()
    conn = get_pymysql_connection()
    cur = conn.cursor()

    records = DataGenerator.generate_simple_tuples(record_count)
    sql = "INSERT INTO simple_records (name, email, age, score) VALUES (%s, %s, %s, %s)"

    cur.executemany(sql, records)
    conn.commit()
    cur.close()
    conn.close()


def bench_pymysql_executemany_batched(record_count: int, batch_size: int = 1000):
    """PyMySQL: executemany with batching"""
    cleanup_table()
    conn = get_pymysql_connection()
    cur = conn.cursor()

    records = DataGenerator.generate_simple_tuples(record_count)
    sql = "INSERT INTO simple_records (name, email, age, score) VALUES (%s, %s, %s, %s)"

    for i in range(0, len(records), batch_size):
        batch = records[i:i + batch_size]
        cur.executemany(sql, batch)

    conn.commit()
    cur.close()
    conn.close()


def bench_pymysql_multi_value(record_count: int, batch_size: int = 1000):
    """PyMySQL: Multi-value INSERT"""
    cleanup_table()
    conn = get_pymysql_connection()
    cur = conn.cursor()

    records = DataGenerator.generate_simple_tuples(record_count)

    for i in range(0, len(records), batch_size):
        batch = records[i:i + batch_size]
        placeholders = ','.join(['(%s, %s, %s, %s)'] * len(batch))
        sql = f"INSERT INTO simple_records (name, email, age, score) VALUES {placeholders}"

        # フラット化
        flat_values = []
        for r in batch:
            flat_values.extend(r)

        cur.execute(sql, flat_values)

    conn.commit()
    cur.close()
    conn.close()


# ========================================
# MySQLdb Benchmarks
# ========================================

def bench_mysqldb_execute(record_count: int):
    """MySQLdb: 1件ずつexecute（遅い）"""
    cleanup_table()
    conn = get_mysqldb_connection()
    cur = conn.cursor()

    records = DataGenerator.generate_simple_tuples(record_count)
    sql = "INSERT INTO simple_records (name, email, age, score) VALUES (%s, %s, %s, %s)"

    for record in records:
        cur.execute(sql, record)

    conn.commit()
    cur.close()
    conn.close()


def bench_mysqldb_executemany(record_count: int):
    """MySQLdb: executemany"""
    cleanup_table()
    conn = get_mysqldb_connection()
    cur = conn.cursor()

    records = DataGenerator.generate_simple_tuples(record_count)
    sql = "INSERT INTO simple_records (name, email, age, score) VALUES (%s, %s, %s, %s)"

    cur.executemany(sql, records)
    conn.commit()
    cur.close()
    conn.close()


def bench_mysqldb_executemany_batched(record_count: int, batch_size: int = 1000):
    """MySQLdb: executemany with batching"""
    cleanup_table()
    conn = get_mysqldb_connection()
    cur = conn.cursor()

    records = DataGenerator.generate_simple_tuples(record_count)
    sql = "INSERT INTO simple_records (name, email, age, score) VALUES (%s, %s, %s, %s)"

    for i in range(0, len(records), batch_size):
        batch = records[i:i + batch_size]
        cur.executemany(sql, batch)

    conn.commit()
    cur.close()
    conn.close()


def bench_mysqldb_multi_value(record_count: int, batch_size: int = 1000):
    """MySQLdb: Multi-value INSERT"""
    cleanup_table()
    conn = get_mysqldb_connection()
    cur = conn.cursor()

    records = DataGenerator.generate_simple_tuples(record_count)

    for i in range(0, len(records), batch_size):
        batch = records[i:i + batch_size]
        placeholders = ','.join(['(%s, %s, %s, %s)'] * len(batch))
        sql = f"INSERT INTO simple_records (name, email, age, score) VALUES {placeholders}"

        flat_values = []
        for r in batch:
            flat_values.extend(r)

        cur.execute(sql, flat_values)

    conn.commit()
    cur.close()
    conn.close()


# ========================================
# LOAD DATA INFILE (最速 - ファイル経由)
# ========================================

def bench_load_data_infile(record_count: int):
    """LOAD DATA LOCAL INFILE を使用（最速）"""
    import tempfile
    import csv

    cleanup_table()

    # 一時CSVファイル作成
    records = DataGenerator.generate_simple_records(record_count)
    with tempfile.NamedTemporaryFile(mode='w', delete=False, suffix='.csv') as f:
        writer = csv.writer(f)
        for r in records:
            writer.writerow([r['name'], r['email'], r['age'], r['score']])
        temp_file = f.name

    try:
        conn = get_pymysql_connection()
        cur = conn.cursor()

        sql = f"""
        LOAD DATA LOCAL INFILE '{temp_file}'
        INTO TABLE simple_records
        FIELDS TERMINATED BY ','
        LINES TERMINATED BY '\n'
        (name, email, age, score)
        """

        cur.execute(sql)
        conn.commit()
        cur.close()
        conn.close()
    finally:
        import os
        os.unlink(temp_file)


def main():
    """メインベンチマーク実行"""
    runner = BenchmarkRunner(output_file='/results/python_mysql_results.json')

    test_sizes = [
        100_000,
        1_000_000,
        # 10_000_000,
    ]

    for size in test_sizes:
        print(f"\n{'#'*60}")
        print(f"# Testing with {size:,} records")
        print(f"{'#'*60}")

        # 1件ずつは遅すぎるので10万件のみ
        if size <= 100_000:
            runner.run_benchmark(
                f"pymysql_execute_{size}",
                bench_pymysql_execute,
                size
            )
            runner.run_benchmark(
                f"mysqldb_execute_{size}",
                bench_mysqldb_execute,
                size
            )

        # PyMySQL
        runner.run_benchmark(
            f"pymysql_executemany_{size}",
            bench_pymysql_executemany,
            size
        )

        runner.run_benchmark(
            f"pymysql_executemany_batched_{size}",
            bench_pymysql_executemany_batched,
            size
        )

        runner.run_benchmark(
            f"pymysql_multi_value_{size}",
            bench_pymysql_multi_value,
            size
        )

        # MySQLdb
        runner.run_benchmark(
            f"mysqldb_executemany_{size}",
            bench_mysqldb_executemany,
            size
        )

        runner.run_benchmark(
            f"mysqldb_executemany_batched_{size}",
            bench_mysqldb_executemany_batched,
            size
        )

        runner.run_benchmark(
            f"mysqldb_multi_value_{size}",
            bench_mysqldb_multi_value,
            size
        )

        # LOAD DATA INFILE
        runner.run_benchmark(
            f"load_data_infile_{size}",
            bench_load_data_infile,
            size
        )

    runner.save_results()
    runner.print_summary()


if __name__ == '__main__':
    print("MySQL Bulk Insert Benchmark")
    print("=" * 60)
    main()
